src/silly_sprites/opengl/buffer_object.cpp

```diff
@@ -0,0 +1,97 @@
+#include "buffer_object.hpp"
+
+namespace sly::gl {
+    void BufferObject::init() {
+        generate_vertex_array();
+        generate_vertex_buffer();
+        generate_element_buffer();
+
+        glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(float), static_cast<void*>(0));
+        glEnableVertexAttribArray(0);
+    }
+
+    void BufferObject::generate_vertex_array() {
+        glGenVertexArrays(1, &m_vao);
+        bind_vertex_array();
+    }
+
+    void BufferObject::generate_vertex_buffer() {
+        glGenBuffers(1, &m_vbo);
+        bind_vertex_buffer();
+        data_vertex_buffer();
+    }
+
+    void BufferObject::generate_element_buffer() {
+        glGenBuffers(1, &m_ebo);
+        bind_element_buffer();
+        data_element_buffer();
+    }
+
+    void BufferObject::bind_vertex_array() const {
+        glBindVertexArray(m_vao);
+    }
+
+    void BufferObject::bind_vertex_buffer() const {
+        glBindBuffer(GL_ARRAY_BUFFER, m_vbo);
+    }
+
+    void BufferObject::bind_element_buffer() const {
+        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, m_ebo);
+    }
+
+    void BufferObject::data_vertex_buffer() const {
+        glBufferData(GL_ARRAY_BUFFER, static_cast<GLsizeiptr>(sizeof(float) * m_points.size()), m_points.data(), GL_STATIC_DRAW);
+        /*
+            ugly cast because of ubuntu.
+            you can see the error here: https://github.com/mgerhold/silly-sprites/actions/runs/5992560779/job/16252072736
+            maybe you have a better idea.
+        */
+    }
+
+    void BufferObject::data_element_buffer() const {
+        glBufferData(GL_ELEMENT_ARRAY_BUFFER, static_cast<GLsizeiptr>(sizeof(unsigned int) * m_indices.size()), m_indices.data(), GL_STATIC_DRAW);
+        // ugly cast again
+    }
+
+    BufferObject::BufferObject(Points points, Indices indices) : m_points{ points }, m_indices{ indices } {
+        init();
+    }
+
+    BufferObject::~BufferObject() {
+        glDeleteBuffers(1, &m_vbo);
+        glDeleteBuffers(1, &m_ebo);
+        glDeleteVertexArrays(1, &m_vao);
+    }
+
+    void BufferObject::bind() const {
+        glBindVertexArray(m_vao);
+    }
+
+    void BufferObject::unbind() const {
+        glBindVertexArray(0);
+    }
+
+    void BufferObject::draw() const {
+        glDrawElements(GL_TRIANGLES, 9, GL_UNSIGNED_INT, 0);
```

Comment:
The `9` is hard-coded. Is this correct?
Since the last parameter is a pointer, you should pass `nullptr` instead of `0` here, I think.

Reply:
how do i manage the index count? should i save this?
right now i would cast the size of the index vector into a GLsizei like so:
`gsl::narrow_cast<GLsizei(m_indeces.size())`
i could not do that when i remove the indeces from the class.

Reply:
I'd suggest to just store the size of the index buffer, but not the index buffer itself (on the CPU side).